[Feature/recruitment serializer] recruitment_serializer test_code 작성 및 Tag 전체적으로 리펙토링

romiii35 opened the pull request:
recruitment_serializer test_code 작성 및 Tag 전체적으로 리펙토링

## ✅ PR 요약
- 작업 요약: recruitment_serializer test_code 작성 및 Tag 전체적으로 리펙토링

## 📄 상세 내용
- [X] 커버리지 상승을 위해 serializer test_code 작성하였습니다.
- [X] create/update/list/detail serializers test 진행
- [X]  tag serializer 수정하였습니다
- [X] tag view, urls 맵핑 및 service 작성하였습니다.
- [X] tags_test 코드 작성하였습니다.

## 🧪 PR Checklist
- [X] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [X] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

romiii35 committed:
✨ feat : recruitment_serializer test_code 작성

recruitment_serializer test_code 작성

romiii35 committed:
💡 chore : tags_serializer 수정

 tags_serializer 수정

romiii35 committed:
✨ feat : tags_view, url 맵핑, 서비스 로직 작성

tags_view, url 맵핑, 서비스 로직 작성

romiii35 committed:
💡 chore : url/init.py 수정(tags 추가)

url/init.py 수정(tags 추가)

romiii35 committed:
✨ feat : tags test_code 작성

tags test_code 작성

romiii35 committed:
💡 chore : isort, black 코드 포멧팅

chore : isort, black 코드 포멧팅

romiii35 committed:
💡 chore : tag test_code 수정

 tag test_code 수정

SWWS97 commented on apps/recruitment/serializers/tags.py:
```diff
@@ -1,26 +1,50 @@
 class TagSerializer(serializers.ModelSerializer[Tag]):
     """태그 조회 및 등록"""
 
+    name = serializers.CharField(required=True, allow_blank=True)
```
여기 required=True -> 필수 줬는데 뒤에 allow_blank=True는 뭐에요?
allow_blank=True 이부분 지워주세요

SWWS97 commented on apps/recruitment/serializers/tags.py:
```diff
@@ -1,26 +1,50 @@
+    def create(self, validated_data: Any) -> Tag:
+        name = validated_data["name"]
+
+        tag, created = TagService.create_tag(name)
```
이부분 지금 tag가 None 이거나 created=false일떄 태그 생성 실패 에러가 발생
발생 할텐데 그러면 이미 존재하는 태그 에러까지 도달 못할거고 의미가 없을거 같아요
tag, created = Tag.objects.get_or_create(name=name)
이렇게 get_or_create를 사용하고 존재하는 태그만 에러만 내보내는 형식으로
작성해주세요

SWWS97 commented on apps/recruitment/serializers/tags.py:
```diff
@@ -30,22 +54,18 @@ class RecruitmentTagUpdateSerializer(serializers.Serializer[Any]):
 
     def validate_tags(self, value: List[int]) -> List[int]:
-
-        # 태그는 최소 1개
```
여기 부분은 지금 위에서 allow_empty=False로 빈 값 못받게 해놨으니 필요 없는 검증이에요

SWWS97 commented on apps/recruitment/services/tags_service.py:
```diff
@@ -0,0 +1,27 @@
+        return qs.order_by("name")
+
+    @staticmethod
+    def create_tag(name: str) -> Tuple[Optional[Tag], bool]:
```
여기 부분 동시성 문제 해결해야할거 같아요 만약에 사용자A,B 두사람이 동시에 태그를 생성하면 IntegrityError가 발생할건데 예외처리 넣어서 이미 생성 됐으면 IntegrityError를 내뱉고 해당 태그를 찾아서 반환 하게끔

romiii35 replied:
serializer에서 원래 TagService 부분을 호출 했었는데,  get_or_create로 바꾸게 된다면.. TagService의 def create_tag은 사용을 안하기 때문에 지우면 될까요..??

SWWS97 commented on apps/recruitment/services/tags_service.py:
```diff
@@ -0,0 +1,27 @@
+        if not name:
+            return None, False
+
+        if Tag.objects.filter(name__iexact=name).exists():  # 존재 여부만 빠르게 확인
```
여기도 존재 여부 확인하는 쿼리를 날리는데 예외처리 넣어서 그냥 tag = Tag.objects.create(name=name) 쿼리 한번 날리고
실패할시에만  Tag.objects.get으로 가져오는걸로 하는게 좋아보여요

romiii35 replied:
위.. 코멘트와 동일합니다.

SWWS97 commented on apps/recruitment/services/tags_service.py:
```diff
@@ -0,0 +1,27 @@
+
+    @staticmethod
+    def create_tag(name: str) -> Tuple[Optional[Tag], bool]:
+        name = (name or "").strip()
```
여기도 소문자든 대문자든 구분을 해야 할거 같습니다 PYTHON, python, Python 이렇게 통일성이 없으니까요

SWWS97 commented on apps/recruitment/services/tags_service.py:
```diff
@@ -0,0 +1,27 @@
+    def get_tags(keyword: str = "") -> QuerySet[Tag]:
+        qs = Tag.objects.all()
+        if keyword:
+            qs = qs.filter(Q(name__iexact=keyword) | Q(name__icontains=keyword))
```
여기도 iexact, icontains 두개가 필요 할거 같진 않아서 qs = qs.filter(name__icontains=keyword) 이렇게 하나만 넣죠

SWWS97 commented on apps/recruitment/tests/test_tags.py:
```diff
@@ -0,0 +1,125 @@
+from django.contrib.auth import get_user_model
+from django.urls import reverse
+from rest_framework import status
+from rest_framework.exceptions import ValidationError
```
안쓰는 import 지워주세요

SWWS97 commented on apps/recruitment/tests/test_tags.py:
```diff
@@ -0,0 +1,125 @@
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertTrue(Tag.objects.filter(name="FastAPI").exists())
+
+    def test_tag_creation_conflict(self) -> None:
```
대소문자일 경우의 대한 테스트도 있으면 좋을거 같아요

SWWS97 commented on apps/recruitment/tests/test_tags.py:
```diff
@@ -0,0 +1,125 @@
+        self.assertIn("name", serializer.errors)
+        self.assertIn("20자 이하", str(serializer.errors["name"]))
+
+    def test_tag_name_invalid_characters(self) -> None:
```
여기서도 하이픈, 공백, 숫자, 자음일경우 허용되는지 테스트도 필요할 거 같습니다

SWWS97 commented on apps/recruitment/tests/test_recruitment_serializer.py:
```diff
@@ -0,0 +1,149 @@
+        self.tag2 = Tag.objects.create(name="Django")
+
+        # 공고 생성용 데이터
+        self.valid_data = {
```
지금 여기도 copy() 해서 사용해서 중복 코드가 좀 보이는거 같은데 헬퍼 메서드로 따로 빼서 data = self._get_valid_recruitment_data(title="제목") 이렇게 가져와서 사용하는게 좋아보여요

SWWS97 commented on apps/recruitment/tests/test_recruitment_serializer.py:
```diff
@@ -0,0 +1,149 @@
+        self.assertIn("lectures", data)
+        self.assertIn("tags", data)
+
+    def test_detail_serializer_output(self) -> None:
```
위의 test_list_serializer_output 이랑 똑같이 테스트용 공고를 생성하는데 위에 setUp부분으로 빼서 미리 생성해서
아래에서 사용하는 형태로 해보죠

romiii35 committed:
💡 chore : 코멘트 참조하여 전반적으로 수정(serializer_tag, test_tags, recruitment_serializer, tags_service)

chore : 코멘트 참조하여 전반적으로 수정(serializer_tag, test_tags, recruitment_serializer, tags_service)

romiii35 committed:
💡 chore : isort, black 코드 포멧팅

 isort, black 코드 포멧팅

romiii35 committed:
💡 chore :  serializer_tags 안쓰는 import 제거, service 로직 변경

 serializer_tags 안쓰는 import 제거, service 로직 변경
